File: platform/polyglot/github_api/py/github_api/sdk/validation.py
"""
Input validation for GitHub API parameters.

Validates repository names, usernames, and branch names against
GitHub's naming rules and reserved names.
"""

from __future__ import annotations

import re
from typing import Any

from github_api.sdk.errors import ValidationError

__all__ = [
    "validate_repository_name",
    "validate_username",
    "validate_branch_name",
    "RESERVED_REPO_NAMES",
]

RESERVED_REPO_NAMES: frozenset[str] = frozenset({
    "settings",
    "security",
    "pulls",
    "issues",
    "actions",
    "apps",
    "codespaces",
    "copilot",
    "discussions",
    "explore",
    "features",
    "marketplace",
    "new",
    "notifications",
    "organizations",
    "packages",
    "projects",
    "search",
    "sponsors",
    "stars",
    "topics",
    "trending",
    "wiki",
    "collections",
    "events",
    "gist",
    "gists",
    "login",
    "logout",
    "pricing",
    "readme",
    "sessions",
    "signup",
    "status",
    "support",
    "terms",
    "watching",
    "account",
    "billing",
    "blog",
    "bounty",
    "case-studies",
    "community",
    "customer-stories",
    "developer",
    "education",
    "enterprise",
    "get-started",
    "graphql",
    "guides",
    "integrations",
    "about",
    "api",
    "docs",
    "join",
    "rest",
})

_REPO_NAME_RE = re.compile(r"^[a-zA-Z0-9._-]+$")
_USERNAME_RE = re.compile(r"^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?$")
_BRANCH_CONTROL_RE = re.compile(r"[\x00-\x1f\x7f]")
_BRANCH_INVALID_CHARS_RE = re.compile(r"[ ~^:?*\[\\]")


def validate_repository_name(name: str) -> str:
    """Validate a GitHub repository name.

    Rules:
    - Non-empty, max 100 characters
    - Only alphanumeric, hyphens, underscores, and dots
    - Cannot start or end with a dot
    - Cannot be a reserved name

    Args:
        name: The repository name to validate.

    Returns:
        The validated repository name.

    Raises:
        ValidationError: If the name violates any rule.
    """
    if not name:
        raise ValidationError(
            "Repository name cannot be empty",
            errors=[{"field": "name", "code": "missing"}],
        )

    if len(name) > 100:
        raise ValidationError(
            f"Repository name exceeds maximum length of 100 characters (got {len(name)})",
            errors=[{"field": "name", "code": "too_long"}],
        )

    if not _REPO_NAME_RE.match(name):
        raise ValidationError(
            f"Repository name contains invalid characters: {name!r}. "
            "Only alphanumeric characters, hyphens, underscores, and dots are allowed.",
            errors=[{"field": "name", "code": "invalid"}],
        )

    if name.startswith(".") or name.endswith("."):
        raise ValidationError(
            f"Repository name cannot start or end with a dot: {name!r}",
            errors=[{"field": "name", "code": "invalid_format"}],
        )

    if name.lower() in RESERVED_REPO_NAMES:
        raise ValidationError(
            f"Repository name {name!r} is reserved by GitHub",
            errors=[{"field": "name", "code": "reserved"}],
        )

    return name


def validate_username(owner: str) -> str:
    """Validate a GitHub username or organization name.

    Rules:
    - Non-empty, max 39 characters
    - Alphanumeric and hyphens only
    - Cannot start or end with a hyphen
    - No consecutive hyphens

    Args:
        owner: The username/org name to validate.

    Returns:
        The validated username.

    Raises:
        ValidationError: If the username violates any rule.
    """
    if not owner:
        raise ValidationError(
            "Username cannot be empty",
            errors=[{"field": "owner", "code": "missing"}],
        )

    if len(owner) > 39:
        raise ValidationError(
            f"Username exceeds maximum length of 39 characters (got {len(owner)})",
            errors=[{"field": "owner", "code": "too_long"}],
        )

    if owner.startswith("-") or owner.endswith("-"):
        raise ValidationError(
            f"Username cannot start or end with a hyphen: {owner!r}",
            errors=[{"field": "owner", "code": "invalid_format"}],
        )

    if "--" in owner:
        raise ValidationError(
            f"Username cannot contain consecutive hyphens: {owner!r}",
            errors=[{"field": "owner", "code": "invalid_format"}],
        )

    if not _USERNAME_RE.match(owner):
        raise ValidationError(
            f"Username contains invalid characters: {owner!r}. "
            "Only alphanumeric characters and single hyphens are allowed.",
            errors=[{"field": "owner", "code": "invalid"}],
        )

    return owner


def validate_branch_name(branch: str) -> str:
    """Validate a Git branch name.

    Rules:
    - Non-empty, max 255 characters
    - No control characters
    - No space, ~, ^, :, ?, *, [, or backslash
    - No consecutive slashes
    - Cannot be a single '@'

    Args:
        branch: The branch name to validate.

    Returns:
        The validated branch name.

    Raises:
        ValidationError: If the branch name violates any rule.
    """
    if not branch:
        raise ValidationError(
            "Branch name cannot be empty",
            errors=[{"field": "branch", "code": "missing"}],
        )

    if len(branch) > 255:
        raise ValidationError(
            f"Branch name exceeds maximum length of 255 characters (got {len(branch)})",
            errors=[{"field": "branch", "code": "too_long"}],
        )

    if _BRANCH_CONTROL_RE.search(branch):
        raise ValidationError(
            f"Branch name contains control characters: {branch!r}",
            errors=[{"field": "branch", "code": "invalid"}],
        )

    if _BRANCH_INVALID_CHARS_RE.search(branch):
        raise ValidationError(
            f"Branch name contains invalid characters: {branch!r}. "
            "Characters space, ~, ^, :, ?, *, [, and \\ are not allowed.",
            errors=[{"field": "branch", "code": "invalid"}],
        )

    if "//" in branch:
        raise ValidationError(
            f"Branch name cannot contain consecutive slashes: {branch!r}",
            errors=[{"field": "branch", "code": "invalid_format"}],
        )

    if branch == "@":
        raise ValidationError(
            "Branch name cannot be a single '@'",
            errors=[{"field": "branch", "code": "invalid_format"}],
        )

    return branch

File: platform/polyglot/github_api/mjs/src/sdk/validation.mjs
/**
 * Input validation for GitHub API operations.
 * Enforces GitHub's naming conventions and reserved name restrictions.
 * @module sdk/validation
 */

import { ValidationError } from './errors.mjs';

/**
 * Reserved repository names that GitHub does not allow.
 * @type {Set<string>}
 */
export const RESERVED_REPO_NAMES = new Set([
  'settings',
  'security',
  'pulls',
  'issues',
  'actions',
  'apps',
  'codespaces',
  'copilot',
  'discussions',
  'explore',
  'features',
  'marketplace',
  'new',
  'notifications',
  'organizations',
  'packages',
  'projects',
  'search',
  'sponsors',
  'stars',
  'topics',
  'trending',
  'wiki',
  'collections',
  'events',
  'gist',
  'gists',
  'login',
  'logout',
  'pricing',
  'readme',
  'sessions',
  'signup',
  'status',
  'support',
  'terms',
  'watching',
  'account',
  'billing',
  'blog',
  'bounty',
  'case-studies',
  'community',
  'customer-stories',
  'developer',
  'education',
  'enterprise',
  'get-started',
  'graphql',
  'guides',
  'integrations',
  'about',
  'api',
  'docs',
  'join',
  'rest',
]);

/** @type {import('./client.mjs').Logger} */
const log = {
  info(msg, ctx) { console.info(`[validation] ${msg}`, ctx || ''); },
  debug(msg, ctx) { console.debug(`[validation] ${msg}`, ctx || ''); },
  warn(msg, ctx) { console.warn(`[validation] ${msg}`, ctx || ''); },
  error(msg, ctx) { console.error(`[validation] ${msg}`, ctx || ''); },
};

/**
 * Validate a GitHub repository name.
 * @param {string} name - The repository name to validate
 * @throws {ValidationError} If the name is invalid
 * @returns {void}
 */
export function validateRepositoryName(name) {
  log.debug('Validating repository name', { name });

  if (!name || typeof name !== 'string') {
    throw new ValidationError('Repository name must be a non-empty string');
  }

  if (name.length > 100) {
    throw new ValidationError(
      `Repository name must be at most 100 characters, got ${name.length}`,
    );
  }

  if (name.startsWith('.') || name.endsWith('.')) {
    throw new ValidationError(
      'Repository name must not start or end with a dot',
    );
  }

  if (!/^[a-zA-Z0-9._-]+$/.test(name)) {
    throw new ValidationError(
      'Repository name may only contain alphanumeric characters, hyphens, underscores, and dots',
    );
  }

  if (RESERVED_REPO_NAMES.has(name.toLowerCase())) {
    throw new ValidationError(
      `Repository name "${name}" is reserved by GitHub`,
    );
  }
}

/**
 * Validate a GitHub username / owner.
 * @param {string} owner - The username to validate
 * @throws {ValidationError} If the username is invalid
 * @returns {void}
 */
export function validateUsername(owner) {
  log.debug('Validating username', { owner });

  if (!owner || typeof owner !== 'string') {
    throw new ValidationError('Username must be a non-empty string');
  }

  if (owner.length > 39) {
    throw new ValidationError(
      `Username must be at most 39 characters, got ${owner.length}`,
    );
  }

  if (owner.startsWith('-') || owner.endsWith('-')) {
    throw new ValidationError('Username must not start or end with a hyphen');
  }

  if (owner.includes('--')) {
    throw new ValidationError('Username must not contain consecutive hyphens');
  }

  if (!/^[a-zA-Z0-9-]+$/.test(owner)) {
    throw new ValidationError(
      'Username may only contain alphanumeric characters and hyphens',
    );
  }
}

/**
 * Validate a Git branch name.
 * @param {string} branch - The branch name to validate
 * @throws {ValidationError} If the branch name is invalid
 * @returns {void}
 */
export function validateBranchName(branch) {
  log.debug('Validating branch name', { branch });

  if (!branch || typeof branch !== 'string') {
    throw new ValidationError('Branch name must be a non-empty string');
  }

  if (branch.length > 255) {
    throw new ValidationError(
      `Branch name must be at most 255 characters, got ${branch.length}`,
    );
  }

  if (branch.includes('//')) {
    throw new ValidationError(
      'Branch name must not contain consecutive slashes',
    );
  }

  if (branch === '@') {
    throw new ValidationError('Branch name must not be a single "@"');
  }

  // eslint-disable-next-line no-control-regex
  if (/[\x00-\x1f\x7f]/.test(branch)) {
    throw new ValidationError('Branch name must not contain control characters');
  }

  if (/[ ~^:?*\[\\]/.test(branch)) {
    throw new ValidationError(
      'Branch name must not contain space, ~, ^, :, ?, *, [, or \\',
    );
  }

  if (branch.endsWith('.lock')) {
    throw new ValidationError('Branch name must not end with ".lock"');
  }

  if (branch.startsWith('.') || branch.endsWith('.')) {
    throw new ValidationError(
      'Branch name must not start or end with a dot',
    );
  }

  if (branch.includes('..')) {
    throw new ValidationError(
      'Branch name must not contain consecutive dots (..)',
    );
  }
}
